### ISSUE TITLE
Incorrect pcc when trying to execute all_gather on 2D tensor

### ISSUE THREAD

**Issue by tapspatel** [tenstorrent/tt-metal#22654]
### Describe the bug

The following is a snippet from multi chip alexnet model.

Working example (pcc = 1.0)
```
import ttnn
import torch

mesh_device = ttnn.open_mesh_device(ttnn.MeshShape(1, 8))

torch_input_tensor = torch.rand((8,2,43264), dtype=torch.float32)

mesh_tensor = ttnn.from_torch(
    torch_input_tensor,
    layout=ttnn.TILE_LAYOUT,
    device=mesh_device,
    mesh_mapper=ttnn.ShardTensor2dMesh(mesh_device, mesh_shape=list(mesh_device.shape), dims=(None, 0)),
)

output_tensor = ttnn.all_gather(
    mesh_tensor,
    dim=0,
    cluster_axis=1,
    mesh_device=mesh_device,
    topology=ttnn.Topology.Linear,
)

torch_tensor = ttnn.to_torch(output_tensor, mesh_composer=ttnn.ConcatMeshToTensor(mesh_device, dim=0))
torch_tensor_first_16 = torch_tensor[:8, :]
get_atol_rtol_pcc(torch_input_tensor, torch_tensor_first_16)

ttnn.visualize_mesh_device(mesh_device, tensor=output_tensor)
ttnn.close_mesh_device(mesh_device)
```

bad example (pcc = 0.18620334189360455)
```
import ttnn
import torch

mesh_device = ttnn.open_mesh_device(ttnn.MeshShape(1, 8))

torch_input_tensor = torch.rand((16,43264), dtype=torch.float32)

mesh_tensor = ttnn.from_torch(
    torch_input_tensor,
    layout=ttnn.TILE_LAYOUT,
    device=mesh_device,
    mesh_mapper=ttnn.ShardTensor2dMesh(mesh_device, mesh_shape=list(mesh_device.shape), dims=(None, 0)),
)

output_tensor = ttnn.all_gather(
    mesh_tensor,
    dim=0,
    cluster_axis=1,
    mesh_device=mesh_device,
    topology=ttnn.Topology.Linear,
)

torch_tensor = ttnn.to_torch(output_tensor, mesh_composer=ttnn.ConcatMeshToTensor(mesh_device, dim=0))
torch_tensor_first_16 = torch_tensor[:16, :]
get_atol_rtol_pcc(torch_input_tensor, torch_tensor_first_16)

ttnn.visualize_mesh_device(mesh_device, tensor=output_tensor)
ttnn.close_mesh_device(mesh_device)
```

Utility function to calculate pcc in comments

**Comment by tapspatel** [comment]
```
def get_atol_rtol_pcc(golden, calculated):
    import numpy as np
    import torch

    # Calculate atol and rtol
    cal_atol = torch.max(torch.abs(golden - calculated)).item()
    cal_rtol = torch.max(torch.abs(golden - calculated) / torch.abs(calculated)).item()

    # Calculate PCC
    def get_pcc(golden, calculated):
        # Both tensors are nan
        if torch.all(torch.isnan(golden)) and torch.all(torch.isnan(calculated)):
            print("Both tensors are 'nan'")
            return 1.0
        # Test if either is completely zero
        elif torch.any(golden.bool()) != torch.any(calculated.bool()):
            return 0.0
        # One tensor is all nan, the other is not
        elif torch.all(torch.isnan(golden)) or torch.all(torch.isnan(calculated)):
            print("One tensor is all nan, the other is not.")
            return 0.0
        else:
            # For now, mask all infs and nans so that we check the rest... TODO
            golden = golden.clone()
            golden[
                torch.logical_or(
                    torch.isnan(golden),
                    torch.logical_or(torch.isinf(golden), torch.isneginf(golden)),
                )
            ] = 0
            calculated = calculated.clone()
            calculated[
                torch.logical_or(
                    torch.isnan(calculated),
                    torch.logical_or(
                        torch.isinf(calculated), torch.isneginf(calculated)
                    ),
                )
            ] = 0

            if torch.equal(golden, calculated):
                return 1.0

            if golden.dtype == torch.bfloat16:
                golden = golden.type(torch.float32)
                calculated = calculated.type(torch.float32)

            # Single element case
            if golden.numel() == 1:
                return float(torch.equal(golden, calculated))

            # If both tensors are contant
            if torch.max(golden) == torch.min(golden) and torch.max(
                calculated
            ) == torch.min(calculated):
                return torch.isclose(torch.max(golden), torch.max(calculated)).item()

            cal_pcc = np.ma.corrcoef(
                np.ma.masked_invalid(torch.squeeze(golden).detach().numpy()).flatten(),
                np.ma.masked_invalid(
                    torch.squeeze(calculated).detach().numpy()
                ).flatten(),
            )
            # Remove correlation coefficient with self (typically always 1.0)
            mask = np.ones(cal_pcc.shape, dtype=bool)
            np.fill_diagonal(mask, 0)
            cal_pcc = np.min(cal_pcc[mask])

            if isinstance(cal_pcc, np.ma.core.MaskedConstant):
                return 1.0

            return cal_pcc

    cal_pcc = get_pcc(golden, calculated)

    print(cal_pcc)

    return (
        cal_atol,
        cal_rtol,
        cal_pcc,
        f"Max ATOL Delta: {cal_atol}, Max RTOL Delta: {cal_rtol}, PCC: {cal_pcc}",
    )
```

**Comment by zzigler-tt** [comment]
@cmaryanTT can you please look into this.

**Comment by cmaryanTT** [comment]
@ntarafdar - beyond fixing whatever the issue is here, can you also add this to the test coverage of CCLs.

**Comment by ajakovljevicTT** [comment]
thank you @tapspatel for opening the issue and to @ntarafdar for following up! Do you have any estimations on when this bug could be resolved, as it is blocking both multichip inference and training currently on the compiler frontend?

**Comment by hkwonTT** [comment]
Perhaps—more precisely—the issue occurs when both of the following conditions are met:
	1.	The dim argument passed in is either the last dimension or the next-to-last (last − 1) dimension.
	2.	The dimension being gathered is not a multiple of 32.

When these two conditions are satisfied, the operation produces incorrect results.

Maybe https://github.com/tenstorrent/tt-metal/issues/21964 is same issue.

**Comment by amorrisonTT** [comment]
Essentially the same problem as https://github.com/tenstorrent/tt-metal/issues/21964

**Comment by amorrisonTT** [comment]
Closed by #30998
